File: mirea/cources/2026-2027/aii-august/08/contest/B.py
mod = 998244353
def solve():
    # читаем данные:
    n = int(input())
    a = list(map(int, input().split()))
    # лестница - комбинация из двух подпоследовательностей: возрастающей и убывающей
    # для каждой позиции считаем количество возрастающих
    # подпоследовательностей, заканчивающихся в ней:
    dpL = [1] * n
    for i in range(1, n):
        for j in range(i):
            if a[j] < a[i]:
                dpL[i] = (dpL[i] + dpL[j]) % mod
    # зеркально для каждой позиции считаем количество убывающих
    # подпоследовательностей, начинающихся в ней:
    dpR = [1] * n
    for i in range(n-2, -1, -1):
        for j in range(i+1, n):
            if a[j] < a[i]:
                dpR[i] = (dpR[i] + dpR[j]) % mod
    # считаем и выводим ответ:
    answ = 0
    for i in range(n):
        answ = (answ + dpL[i] * dpR[i]) % mod
    print(answ)
tt = int(input())
for _ in range(tt):
    solve()

File: mirea/cources/2026-2027/aii-august/08/contest/B.cpp
#include <bits/stdc++.h>
using namespace std;
const int mod = 998244353;
void solve() {
    // читаем данные:
    int n; cin >> n;
    vector<int> a(n);
    for (auto &it : a) cin >> it;
    // лестница - комбинация из двух подпоследовательностей: возрастающей и убывающей
    // для каждой позиции считаем количество возрастающих
    // подпоследовательностей, заканчивающихся в ней:
    vector<int> dpL(n, 1);
    for (int i = 1; i < n; i++)
        for (int j = 0; j < i; j++)
            if (a[j] < a[i])
                (dpL[i] += dpL[j]) %= mod;
    // зеркально для каждой позиции считаем количество убывающих
    // подпоследовательностей, начинающихся в ней:
    vector<int> dpR(n, 1);
    for (int i = n-2; i >= 0; i--)
        for (int j = i+1; j < n; j++)
            if (a[j] < a[i])
                (dpR[i] += dpR[j]) %= mod;
    // считаем и выводим ответ:
    int64_t answ = 0;
    for (int i = 0; i < n; i++)
        (answ += dpL[i] * 1LL * dpR[i]) %= mod;
    cout << answ << '\n';
}
main() {
    int tt; cin >> tt;
    while (tt --> 0) solve();
}
